shopinvader_dropoff_site/services/cart.py

```diff
@@ -0,0 +1,97 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 Akretion (http://www.akretion.com).
+# @author Sébastien BEAU <sebastien.beau@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+
+from odoo.addons.component.core import Component
+from odoo.exceptions import UserError
+from odoo.tools.translate import _
+
+
+class CartService(Component):
+    _inherit = "shopinvader.cart.service"
+
+    def apply_dropoff_site(self, **params):
+        """
+            This service will apply the given dropoffsite to the current
+            cart
+        :param params: Dict containing dropoff site information
+        :return:
+        """
+        cart = self._get()
+        if not cart:
+            raise UserError(_("There is not cart"))
+        else:
+            self._set_dropoff_site(cart, params)
+            return self._to_json(cart)
+
+    # Validator
+    def _validator_apply_dropoff_site(self):
+        return {
+            "ref": {"type": "string", "required": True},
+            "name": {"type": "string", "required": True},
+            "street": {"type": "string"},
+            "street2": {"type": "string"},
+            "zip": {"type": "string", "required": True},
+            "city": {"type": "string", "required": True},
+            "phone": {"type": "string"},
+            "state_code": {"type": "string"},
+            "country_code": {"type": "string", "required": True},
+        }
+
+    def _prepare_dropoff_site_params(self, cart, dropoff_site):
+        country_code = dropoff_site.pop("country_code")
+        state_code = dropoff_site.pop("state_code", None)
+
+        country = self.env["res.country"].search([("code", "=", country_code)])
+        if not country:
+            raise UserError(_("Invalid country code %s") % country_code)
+        dropoff_site["country_id"] = country.id
+
+        if state_code:
+            state = self.env["res.country.state"].search(
+                [("code", "=", state_code), ("country_id", "=", country.id)]
+            )
+            if not state:
+                raise UserError(
+                    _("Invalid state code %s for country %s")
+                    % (state_code, country_code)
+                )
+            dropoff_site["state_id"] = state.id
+
+        if not cart.carrier_id:
+            raise UserError(_("You must select a carrier first"))
+        dropoff_site["carrier_id"] = cart.carrier_id.id
+        return dropoff_site
+
+    def _set_dropoff_site(self, cart, dropoff_site):
+        vals = self._prepare_dropoff_site_params(cart, dropoff_site)
+        dropoff_site_obj = self.env["dropoff.site"]
+        dropoff_site = dropoff_site_obj.search(
+            [
+                ("carrier_id", "=", cart.carrier_id.id),
+                ("ref", "=", dropoff_site["ref"]),
+            ]
+        )
+        if dropoff_site:
+            dropoff_site.write(vals)
+        else:
+            dropoff_site = dropoff_site_obj.create(vals)
```

Comment:
@sebastienbeau I'm a little bit surprised to see here the creation of dropoff site. I don't like the idea that the same method could retrieve and apply an existing dropoff site or create a new one.  In each cases, dropoff-site should be created into the backend event if that means that the backend must call the same external api as the site to get the information.  With such an implementation the apply_dropoff_site/set_dropoff_size would take only an identifier as parameter (for some carrier, this identifier will be the odoo one and in other case, the carrier one...)

IMO, we should also provide a new service endpoint 'dropff.service' that could be use to search drop-off site.  With such a service, we could use this new shopinvader addon to expose dropoff site manually managed into Odoo.

Reply:
I think also that having public method that fill all dropoffsite information is not really clean.
I have splited the module because the base module should not have such feature.
Sadly it seem that it's the way to do if we want to use external widget like laposte.
So for now I have done 2 new module shopinvader_dropoff_site_edition and shopinvader_dropoff_site_laposte.

I also have renamed the method and add a todo for the search part